pkg/blobstore/replication/metrics_blob_replicator.go

```diff
@@ -0,0 +1,165 @@
+package replication
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/buildbarn/bb-storage/pkg/blobstore/buffer"
+	"github.com/buildbarn/bb-storage/pkg/blobstore/slicing"
+	"github.com/buildbarn/bb-storage/pkg/clock"
+	"github.com/buildbarn/bb-storage/pkg/digest"
+	"github.com/buildbarn/bb-storage/pkg/util"
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var (
+	replicatorOperationsPrometheusMetrics sync.Once
+
+	replicatorOperationsDurationSeconds = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "buildbarn",
+			Subsystem: "blobstore",
+			Name:      "replicator_operations_duration_seconds",
+			Help:      "Amount of time spent per operation on blob replicator, in seconds.",
+			Buckets:   util.DecimalExponentialBuckets(-3, 6, 2),
+		},
+		[]string{"operation"})
+
+	replicatorOperationsBlobSizeBytes = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "buildbarn",
+			Subsystem: "blobstore",
+			Name:      "replicator_operations_blob_size_bytes",
+			Help:      "Size of blobs being replicated, in bytes.",
+			Buckets:   prometheus.ExponentialBuckets(1.0, 2.0, 33),
+		},
+		[]string{"operation"})
+
+	replicatorOperationsBatchSize = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "buildbarn",
+			Subsystem: "blobstore",
+			Name:      "replicator_operations_batch_size",
+			Help:      "Number of blobs in batch replication requests.",
+			Buckets:   prometheus.ExponentialBuckets(1.0, 2.0, 17),
+		},
+		[]string{"operation"})
+)
+
+type metricsBlobReplicator struct {
+	replicator  BlobReplicator
+	clock       clock.Clock
+	source      string
+	destination string
+
+	singleDurationSeconds    prometheus.ObserverVec
+	singleBlobSizeBytes      prometheus.Observer
+	compositeDurationSeconds prometheus.ObserverVec
+	compositeBlobSizeBytes   prometheus.Observer
+	multipleDurationSeconds  prometheus.ObserverVec
+	multipleBatchSize        prometheus.Observer
+	multipleBlobSizeBytes    prometheus.Observer
+}
+
+// NewMetricsBlobReplicator creates a wrapper around BlobReplicator that adds
+// Prometheus metrics for monitoring replication operations.
+func NewMetricsBlobReplicator(replicator BlobReplicator, clock clock.Clock) BlobReplicator {
```

Comment:
Right now these metrics aren't broken down by backend/storage type. Is that intentional? If not, maybe we can simply use the backend type of the BlobAccess that they are used in conjunction with?

Reply:
added storage type, I didn't see a good way to extract backendType from blobAccess though.